### PR TITLE
fix: allow start charging on a reservation

### PR DESCRIPTION
**Pull request by nahueld**
Although documentation on OCPP 1.6 does not show a transition from Reserved to Charging in the transitions table it also states in the Reserve Now section:
> If the Charge Point accepts the reservation request, then it SHALL refuse charging for all incoming idTags on the reserved > connector, except when the incoming idTag or the parent idTag match the idTag or parent idTag of the reservation.

With this change the simulator will
- Allow start a transaction if the status is `Reserved`
- If there is a idTag mismatch the transaction is stopped

### CODE REVIEW

**Comment by JulianHBuecher** [comment]
Hi @nahueld, 
thanks a lot for your contribution to the existing reservation handling in the simulator. But, the status of the charging station never is `Reserved`, when it reaches this code. After passing the `handleRequestRemoteStart` method in `OCPP16IncomingRequestService`, the `sendAndSetConnectorStatus` method will transition the state from `Reserved` to `Preparing`. So this check is unnecessary.

In case you want to try it, you can utilize my forks of Open e-Mobility implementing the OCPP `ReserveNow` and `CancelReservation`:
- Backend: https://github.com/JulianHBuecher/ev-server/tree/reservation-process-wip
- Frontend: https://github.com/JulianHBuecher/ev-dashboard/tree/reservation-process-wip

**Comment by jerome-benoit** [comment]
> Hi @nahueld, thanks a lot for your contribution to the existing reservation handling in the simulator. But, the status of the charging station never is `Reserved`, when it reaches this code. 

And in case of a local `StartTransaction` from UI services?

File: src/charging-station/ocpp/1.6/OCPP16ResponseService.ts [hunk 1]
```diff
@@ -693,15 +701,19 @@ export class OCPP16ResponseService extends OCPPResponseService {
         }
       }
     }
+
     if (
       connectorStatus?.status !== OCPP16ChargePointStatus.Available &&
-      connectorStatus?.status !== OCPP16ChargePointStatus.Preparing
+      connectorStatus?.status !== OCPP16ChargePointStatus.Preparing &&
+      connectorStatus?.status !== OCPP16ChargePointStatus.Reserved
```
Comment (jerome-benoit) [inline src/charging-station/ocpp/1.6/OCPP16ResponseService.ts:708]: I think the issue is more related to the fact that sending a `StartTransaction` request after at local badge scan simulation is not putting the connector status in preparing state. And if the `StartTransaction` is refused server-side, the previous connector status is not properly restored if a valid reservation is found.  

Would you redo that PR to implement it?

Comment (JulianHBuecher) [reply src/charging-station/ocpp/1.6/OCPP16ResponseService.ts:708]: Hi Jerome, found some time to test it with the UI. 

Regarding your first point:
- sending a StartTransaction request after at local badge scan simulation is not putting the connector status in preparing state
- Behaviour: Absolutely correct. If the connector is reserved, then the code breaks here.

Regarding your second point:
- if the StartTransaction is refused server-side, the previous connector status is not properly restored if a valid reservation is found
- Behaviour: Exactly, if the `StartTransaction` fails, the connector goes back to the `Available` state.

Comment (jerome-benoit) [reply src/charging-station/ocpp/1.6/OCPP16ResponseService.ts:708]: > * sending a StartTransaction request after at local badge scan simulation is not putting the connector status in preparing state
> * Behaviour: Absolutely correct. If the connector is reserved, then the code breaks here.

The OCPP 1.6 specs are not clear about the connector preparing state sent at issuing a `StartTransaction`. But given the fact that: 
- `Reserved` to `Charging` is not a documented connector status transition (and the simulator code checking status transition will warn), 
- Putting a connector into `Preparing` status at `StartTransaction` sounds reasonable even if it's not the strict definition of preparing state in the specs,

the correct fix is to send the `Preparing` status at `StartTransaction` with code review for duplicate sending.

Comment (JulianHBuecher) [reply src/charging-station/ocpp/1.6/OCPP16ResponseService.ts:708]: Yeah, I totally agree here. Using the preparing state at remote start and in the local start makes the overall behavior of the simulator more predictable and easier to understand.

**Pull request closed without merging**